### ISSUE TITLE
Replace Image Registry

### ISSUE THREAD

**Issue by paediatbalcabch** [kyverno/policies#173]
I have created the following policy to replace the image registry:

<details>
  <summary>replace-image-registry-policy</summary>

```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: replace-image-registry
  annotations:
    policies.kyverno.io/title: Replace Image Registry
    policies.kyverno.io/category: Sample
    policies.kyverno.io/severity: medium
    policies.kyverno.io/subject: Pod
    policies.kyverno.io/minversion: 1.3.6
    policies.kyverno.io/description: >-
      Rather than blocking Pods which come from outside registries,
      it is also possible to mutate them so the pulls are directed to
      approved registries. In some cases, those registries may function as
      pull-through proxies and can fetch the image if not cached.
      This policy policy mutates all images either
      in the form 'image:tag' or 'registry.corp.com/image:tag' to be prefaced
      with `my.repo.com:1234/`.
spec:
  background: false
  rules:
    - name: replace-image-registry-containers
      match:
        resources:
          kinds:
          - Pod
      mutate:
        patchStrategicMerge:
          spec:
            containers:
            - (name): "*"
              image: |-
                                {{ regex_replace_all( '^((\w+\.)+\w+(:\d+)?\/)?', '{{@}}', 'my.repo.com:1234/') }}
    - name: replace-image-registry-initcontainers
      match:
        resources:
          kinds:
          - Pod
      mutate:
        patchStrategicMerge:
          spec:
            initContainers:
            - (name): "*"
              image: |-
                                {{ regex_replace_all( '^((\w+\.)+\w+(:\d+)?\/)?', '{{@}}', 'my.repo.com:1234/') }}
```
</details>

but when I apply the following manifest:

<details>
  <summary>manifest with init containers</summary>

```yaml
---
# Source: renku/charts/keycloak/templates/statefulset.yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: renku-keycloak
  labels:
    helm.sh/chart: keycloak-9.8.1
    app.kubernetes.io/name: keycloak
    app.kubernetes.io/instance: renku
    app.kubernetes.io/version: "11.0.2"
    app.kubernetes.io/managed-by: Helm
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: keycloak
      app.kubernetes.io/instance: renku
  replicas: 1
  serviceName: renku-keycloak-headless
  podManagementPolicy: Parallel
  updateStrategy:
    type: RollingUpdate
  template:
    metadata:
      annotations:
        checksum/config-startup: 8075f602d36f32b0c87efa90c53e37bc2f903e945ed1fbf20b496a836b42e9ec
        checksum/secrets: 44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
      labels:
        app.kubernetes.io/name: keycloak
        app.kubernetes.io/instance: renku
    spec:
      initContainers:
        - name: init-certificates
          securityContext:
            allowPrivilegeEscalation: false
            runAsGroup: 65534
            runAsUser: 65534
          image: "redhat/ubi8:8.4"
          command: ["sh", "-c"]
          args: ["mkdir -p /etc/pki/ca-trust/extracted/openssl/ /etc/pki/ca-trust/extracted/pem/ /etc/pki/ca-trust/extracted/java/ /etc/pki/ca-trust/extracted/edk2 && update-ca-trust"]
          volumeMounts:
            - name: etc-ssl-certs
              mountPath: /etc/pki/ca-trust/extracted
            - name: custom-ca-certs
              mountPath: /usr/share/pki/ca-trust-source/anchors
              readOnly: true
        - name: theme-provider
          image: renku/keycloak-theme:2.0.1
          imagePullPolicy: IfNotPresent
          command:
            - sh
          args:
            - -c
            - |
              echo "Copying theme..."
              cp -Rfv /renku_theme/* /theme
          volumeMounts:
            - name: theme
              mountPath: /theme
      containers:
        - name: keycloak
          securityContext:
            runAsNonRoot: true
            runAsUser: 1000
          image: "docker.io/jboss/keycloak:11.0.2"
          imagePullPolicy: IfNotPresent
          command:
            []
          args:
            []
          env:
            - name: DB_VENDOR
              value: postgres
            - name: DB_PORT
              value: "5432"
            - name:  PROXY_ADDRESS_FORWARDING
              value: "true"
            
          envFrom:
            - secretRef:
                name: renku-keycloak-postgres
            - secretRef:
                name: keycloak-password-secret
            
          ports:
            - name: http
              containerPort: 8080
              protocol: TCP
            - name: https
              containerPort: 8443
              protocol: TCP
            - name: http-management
              containerPort: 9990
              protocol: TCP
          livenessProbe:
            httpGet:
              path: /auth/
              port: http
            initialDelaySeconds: 300
            timeoutSeconds: 5
            
          readinessProbe:
            httpGet:
              path: /auth/realms/master
              port: http
            initialDelaySeconds: 30
            timeoutSeconds: 1
            
          resources:
            {}
          volumeMounts:
            - name: startup
              mountPath: "/opt/jboss/startup-scripts/keycloak.cli"
              subPath: "keycloak.cli"
              readOnly: true
            - name: theme
              mountPath: /opt/jboss/keycloak/themes/renku-theme
            - name: etc-ssl-certs
              mountPath: /etc/pki/ca-trust/extracted
              readOnly: true
            
      serviceAccountName: renku-keycloak
      securityContext:
        fsGroup: 1000
      enableServiceLinks: true
      restartPolicy: Always
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchLabels:
                  app.kubernetes.io/name: keycloak
                  app.kubernetes.io/instance: renku
                matchExpressions:
                  - key: app.kubernetes.io/component
                    operator: NotIn
                    values:
                      - test
              topologyKey: kubernetes.io/hostname
          preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 100
              podAffinityTerm:
                labelSelector:
                  matchLabels:
                    app.kubernetes.io/name: keycloak
                    app.kubernetes.io/instance: renku
                  matchExpressions:
                    - key: app.kubernetes.io/component
                      operator: NotIn
                      values:
                        - test
                topologyKey: failure-domain.beta.kubernetes.io/zone
        
      terminationGracePeriodSeconds: 60
      volumes:
        - name: startup
          configMap:
            name: renku-keycloak-startup
            defaultMode: 0555
            items:
              - key: keycloak.cli
                path: keycloak.cli
        - name: theme
          emptyDir: {}
        - name: etc-ssl-certs
          emptyDir:
            medium: "Memory"
        - name: custom-ca-certs
          projected:
            defaultMode: 0777
            sources:
              - secret:
                  name: renku-tls
```
</details>

then all the **initContainers** images will be set to **redhat/ubi8:8.4**

my expected behaviour would be, that the the replacement works independently for all initContainers.

**Comment by chipzoller** [comment]
We currently have an issue with the way Kyverno runs this internally and is covered [here](https://github.com/kyverno/kyverno/issues/2243). Looking to get this fixed by 1.6.0.

**Comment by chipzoller** [comment]
This is now resolved and available in an updated policy: https://github.com/kyverno/policies/tree/main/other/replace_image_registry
